feat: support reading/loading multiple .env files

duarte-pompeu opened the pull request:
This PR allows reading from multiple `.env` files in the dotenv CLI, specifically the targets `get`, `list` and `run`.

This will allow users to have a big `.env` file with default configs, and smaller ones which only incremental changes - therefore removing the burden of repeating unchanged key-values.

The goal for the new behavior is as follows:
- files are processed in order
- new keys lead to creating new key-values
- repeated keys lead to overriding previous values

Extra remarks:
- this PR preserves the behavior of working with single files, and falling back to `.env` when no file is specified
- this PR forbids write operations on multiple files, specifically `set` and `unset`

Example:

```shell
$ echo -e 'URL="https://www.example.com"\nUSER="user"\nPASSWORD="secret"' > .env1
$ echo -e 'USER="admin"\nPASSWORD="admin"' > .env2
$ python -m dotenv -f .env1 -f .env2 list
PASSWORD=admin
URL=https://www.example.com
USER=admin
```

Ideas for the future:
- also allow loading multiple `.env` files from the code, eg `load_dotenv(f1, f2, ...)`
- also allow setting and unsetting values for multiple `.env` files

Relates to #418 .

duarte-pompeu commented on src/dotenv/cli.py:
```diff
@@ -44,56 +48,70 @@ def cli(ctx: click.Context, file: Any, quote: Any, export: Any) -> None:
+    if set:
+        click.echo(value)
     else:
+        logger.warning(f"Key {key} not found in {files[0] if len(files) == 1 else files}.")
```
Moved warning to cli: key may not be present in one file, but present in another.

Not sure how this will affect other calls to this function.

duarte-pompeu replied:
Looks a bit strange right now:

```
python3 dotenv/ -f test1.env -f test2.env get URLa
Key URLa not found in test1.env.
Key URLa not found in test2.env.
Key URLa not found in ('test1.env', 'test2.env').
```

duarte-pompeu replied:
Fixed by [f730c6d](https://github.com/theskumar/python-dotenv/pull/424/commits/f730c6d92900f8741ae03f24e177b1a3ef32ecfc)

duarte-pompeu commented on src/dotenv/cli.py:
```diff
@@ -44,56 +48,70 @@ def cli(ctx: click.Context, file: Any, quote: Any, export: Any) -> None:
-        )
-    dotenv_as_dict = dotenv_values(file)
+    
+    d = {}
```
Bad name, I guess.

duarte-pompeu commented on src/dotenv/cli.py:
```diff
@@ -121,22 +145,28 @@ def unset(ctx: click.Context, key: Any) -> None:
-    }
-
+    
+    d = {}
```
Bad name, I guess.

duarte-pompeu commented on tests/test_main.py:
```diff
@@ -87,21 +86,14 @@ def test_get_key_no_file(tmp_path):
             mock.call("Python-dotenv could not find configuration file %s.", nx_file)
         ],
     )
-    mock_warning.assert_has_calls(
-        calls=[
-            mock.call("Key %s not found in %s.", "foo", nx_file)
-        ],
-    )
```
Since I removed the warning, I also removed this mock.

duarte-pompeu replied:
I re-added the warning (by setting DotEnv(...verbose=True).

duarte-pompeu reviewed (commented):
could be improved

duarte-pompeu commented on src/dotenv/cli.py:
```diff
@@ -102,11 +113,17 @@ def get(ctx: click.Context, key: Any) -> None:
+            global_success = True
+            success_files.append(file)
+    if global_success:
+        click.echo("Successfully removed %s from %s" % (key, success_files[0] if len(success_files) == 1 else success_files))
```
this line has a formatting issue but I don't know how to fix it 🤔

duarte-pompeu replied:
Fixed by [2b3a28b](https://github.com/theskumar/python-dotenv/pull/424/commits/2b3a28be6ee47c6e4780c5df6196642bf2e5e894)

duarte-pompeu reviewed (commented):
double checked

duarte-pompeu commented on src/dotenv/main.py:
```diff
@@ -111,13 +111,14 @@ def get_key(
     key_to_get: str,
     encoding: Optional[str] = "utf-8",
+    verbose: bool = True,
 ) -> Optional[str]:
     """
     Get the value of a given key from the given .env.
 
     Returns `None` if the key isn't found or doesn't have a value.
     """
-    return DotEnv(dotenv_path, verbose=True, encoding=encoding).get(key_to_get)
+    return DotEnv(dotenv_path, verbose=verbose, encoding=encoding).get(key_to_get)
```
Added this flag because when invoked from `cli`, we may check multiple files, and if one includes the key, we don't want to emit n warnings.

Therefore the cli can call it with `verbose = False` and check whether it should warn or not.

duarte-pompeu replied:
Doesn't seem needed anymore, as `cli.get` doesn't invoke this anymore.

duarte-pompeu commented:
I can give it a try tonight, it would be interesting to merge this feature.

duarte-pompeu commented:
Solved conflicts but there's still some work to do regarding failing tests.

pheanex commented:
Awesome! Let's see if @bbc2 or @theskumar is available for a quick review?

theskumar commented:
Fundamentally as feature I'm good with supporting this. I like your approach and seeing how you are going about it. I haven't had time to look into the implementation details. But I feel free to improve on your implementation. Would really appreciate keeping it backward compatible. Thank you @duarte-pompeu for your contribution.

theskumar commented:
Read is completely fine, I could not see how the write be extended to multiple files. Would be curious if you have some ideas around it.

duarte-pompeu commented:
> Read is completely fine, I could not see how the write be extended to multiple files. Would be curious if you have some ideas around it.

Actually I already implemented it: `unset` looks for the key in all files and deletes it when present. But I have some doubts about this usefulness, or if this behavior is intuitive to users - as such, I'd rather remove it from this PR and discuss it later.

Example as of [49c34a5](https://github.com/theskumar/python-dotenv/pull/424/commits/49c34a55f3d0929837a2b5b857562af07669478d):

```bash
$ for f in $(ls .env*); do echo $f; cat $f; echo; done
.env1
a=1
b=2
.env2
b=2
c=3
# unset 'a', which is only present in .env1
$ python -m dotenv -f .env1 -f .env2 unset a
Successfully removed a from .env1.
$(ls .env*); do echo $f; cat $f; echo; done
.env1
b=2
.env2
b=2
c=3
# unset 'b', which is present in both .env1 and .env2
$ python -m dotenv -f .env1 -f .env2 unset b
Successfully removed b from ['.env1', '.env2'].
$for f in (ls .env*); do echo $f; cat $f; echo; done
.env1
.env2
c=3
# try to unset key which does not exist
$ python -m dotenv -f .env1 -f .env2 unset d
Key d not removed from ('.env1', '.env2') - key doesn't exist.
```

duarte-pompeu commented on tests/test_cli.py:
```diff
@@ -37,6 +37,25 @@ def test_list(cli, dotenv_path, format: Optional[str], content: str, expected: s
+    dotenv_path.write_text(contents[0] + '\n')
+    extra_dotenv_path.write_text(contents[1] + '\n')
+
+    args = ['--file', dotenv_path, '--file', extra_dotenv_path, 'list']
```
I omitted tests with multiple formatting because that seems to me like a separate concern - if the variables are loaded correctly, the formatting behavior shouldn't change.

Do you agree?

theskumar replied:
You are right. This is fine.

duarte-pompeu commented:
I think this is ready for review @theskumar:

- implemented changes
- kept backward compatibility (I think so, someone please double check for peace of mind 😃)
- added tests
- updated docs
- improved PR description

We didn't talk about this, but I only implemented this for the CLI utility. If there's interest to also enable this behavior from the library code, I'm up to contribute to it (but would prefer to have it in a separate PR, as I find this one already useful as is).

duarte-pompeu commented:
Hi @theskumar, any feedback?

duarte-pompeu commented:
Hi @theskumar and @pheanex , any feedback?

pheanex commented:
Lgtm. Looking forward to this ☺️👍

theskumar commented:
@duarte-pompeu thanks for all the hard work. Unfortunately, I've had not been able to review this yet. Looks like we Christmas 🎄 break would be a good time to spend sometime taking care of it.

duarte-pompeu commented:
> @duarte-pompeu thanks for all the hard work. Unfortunately, I've had not been able to review this yet. Looks like we Christmas 🎄 break would be a good time to taking care of it for me, along with few more pull requests.

Cool, I also have some free time now so it would also work for me 😃

theskumar commented on README.md:
````diff
@@ -154,6 +154,17 @@ $ dotenv list --format=json
+a=1
+b=20
+c=30
+```
````
@duarte-pompeu Consider document the precedence/loading order values from the of the files. I didn't see a test but based on the this example it seems the `.env2` overrides the previous value from `.env1`.

duarte-pompeu replied:
I agree: it's important to document this. I chose an overriding behavior, as my use case would be:

- have a big file with all values
- have smaller .env files with specific values to override in certain situations

I tried to improve it on 363aab5b3b228e670c09019b134570949faf5de3

theskumar commented on README.md:
```diff
@@ -3,7 +3,7 @@
 [![PyPI version][pypi_badge]][pypi_link]
 
-Python-dotenv reads key-value pairs from a `.env` file and can set them as environment
+Python-dotenv reads key-value pairs from `.env` files and can set them as environment
```
I'm hesitant to change this yet as the core/python API, doesn't support reading from multiple files.

duarte-pompeu replied:
Good point. I agree it makes sense to only document this in the CLI section.

Resolved in 1d33b6ad8e09f8bd1ffef73def57b581eafae4b9

duarte-pompeu commented:
Hi @theskumar , I agree with your feedback and addressed in the last commits. What do you think?

pheanex commented:
@theskumar any chance to take a quick look again?

pheanex commented:
@bbc2 any chance you can take a look at this? It seems @theskumar is currently unavailable.

bbc2 reviewed (approved):
I'm OK with this (we might be opening Pandora's box but so be it) and it looks like you addressed all the concerns raised previously, so if no one objects for a few more days I'll go ahead and merge it. Thank you for your contribution!

pheanex commented:
@bbc2 as two more weeks have passed: Are we good to move forward with this?

leodrivera commented:
It seems to be a great feature. Any chances to it get merged?

duarte-pompeu commented:
This is supported by dotenvx, which seems a better way to load dotenvs using the CLI instead of a library call.

It's been 2 years since opening the PR, I'm going to close it.
